idx/metric_definition.go

```diff
@@ -0,0 +1,552 @@
+package idx
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/binary"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+	"github.com/tinylib/msgp/msgp"
+)
+
+//go:generate msgp
+
+func init() {
+	msgp.RegisterExtension(95, func() msgp.Extension { return &MetricName{} })
+	msgp.RegisterExtension(90, func() msgp.Extension { return &TagKeyValues{} })
+}
+
+type mType uint8
+
+// MType represented by a uint8
+const (
+	MTypeUndefined mType = iota
+	MTypeGauge
+	MTypeRate
+	MTypeCount
+	MTypeCounter
+	MTypeTimestamp
+)
+
+// MetricName stores the name as a []uintptr to strings interned in an object store.
+// Each word is stored as a separate interned string without the '.'. The whole name
+// can be retrieved by calling MetricName.String()
+type MetricName struct {
+	nodes []uintptr
+}
+
+// Nodes returns the []uintptr of interned string addresses
+// for the MetricName
+func (mn *MetricName) Nodes() []uintptr {
+	return mn.nodes
+}
+
+// String returns the full MetricName as a string
+// using data interned in the object store
+func (mn *MetricName) String() string {
+	if len(mn.nodes) == 0 {
+		return ""
+	}
+
+	var bld strings.Builder
+	return mn.string(&bld)
+}
+
+func (mn *MetricName) string(bld *strings.Builder) string {
+	name, err := IdxIntern.JoinStrings(mn.nodes, ".")
+	if err != nil {
+		bld.WriteString("invalid")
```

Comment:
should this maybe do `bld.Reset()` first? otherwise the resulting string might end up being whatever nodes have been processed before the error occurred, and then `"invalid"`.

Reply:
No, it will already return "" if there is an error from IdxIntern.JoinStrings